store/src/java/com/zimbra/cs/ml/classifier/ClassifierRegistry.java

```diff
@@ -0,0 +1,197 @@
+package com.zimbra.cs.ml.classifier;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.util.BEncoding;
+import com.zimbra.common.util.BEncoding.BEncodingException;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.mailbox.Message;
+import com.zimbra.cs.ml.Classifiable;
+import com.zimbra.cs.ml.classifier.Classifier.ClassifiableType;
+import com.zimbra.cs.ml.feature.FeatureSet;
+import com.zimbra.cs.ml.feature.FeatureSpec;
+import com.zimbra.cs.ml.schema.ClassifierInfo;
+
+/**
+ * Registry of classifiers known to be initialized on the ML server.
+ * Subclasses handle persisting/loading classifier info.
+ */
+public abstract class ClassifierRegistry {
+
+    private static final String KEY_ID = "i";
+    private static final String KEY_LABEL = "l";
+    private static final String KEY_FEATURES = "f";
+    private static final String KEY_DESCRIPTION = "d";
+    private static final String KEY_TYPE = "t";
+
+    private Map<String, Classifier<?>> idMap;
+    private Map<String, Classifier<?>> labelMap;
+    private boolean loaded = false;
+
+    public ClassifierRegistry() {
+        idMap = new HashMap<>();
+        labelMap = new HashMap<>();
+    }
+
+    void loadKnownClassifiers() throws ServiceException {
```

Comment:
Does this need to be synchronized?  Seem to be a few calls to it.

Reply:
yeah - good catch